# Whitney Hardy 
## Task 4.8 Groupong Data and Aggregating Variables

### Content List
- Import libraries
- Import data
- Aggregate data
- Analyze the results
- Create loyalty flag
- Group data
- Create frequency flag
- Export Data

### Import libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

### Import Data

In [2]:
#create path 
path = r'C:\Users\hardy\10-2025 Instacart Basket Analysis' 

In [3]:
#test path 
path 

'C:\\Users\\hardy\\10-2025 Instacart Basket Analysis'

In [4]:
#import dataframe 
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 data', 'prepared data', 'ords_prods_merge_updated.pkl')) 

In [5]:
#check output
ords_prods_merge.shape

(32399732, 20)

In [6]:
#check output 
ords_prods_merge. head(5) 

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered _merge                             product_name  aisle_id  \
0          0   both                                     Soda        77   
1          0   both  Organic Unsweetened Vanilla Almond Milk        91   
2          0   both                      Original Beef Jerky        23   
3          0   both               Aged White Cheddar Popcorn        23   
4          0   both         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices Match    price_range_loc     busiest_day  \
0              7     9.0  both  Mid-range product  Regularly busy   
1             16    12.5  both  Mid-range product  Regularly busy   
2             19     4.4  both  Low-range product  Regularly busy   
3             19     4.7  both  Low-range product  Regularly busy   
4             17     1.0  both  Low-range product  Regularly busy   

     busiest_days busiest_period_of_day  
0  Regularly busy        Average orders  
1  Regularly busy        Average orders  
2  Regularly busy        Average orders  
3  Regularly busy        Average orders  
4  Regularly busy        Average orders

#### Question 
- I am curios why my user_id repeats so many time. It appears to be exery item in an order is listed but that doesnt seem to be the case in the examples in the readings. Any insight to this would be greatly appreciated as I want to be doing this correctly.

#### Aggregate data

In [8]:
#step 2- aggregate mean of order_number grouped by department_id for entire dataframe
#grouping data using pandas
ords_prods_merge.groupby('product_name') 

In [9]:
#split the data and apply the agg
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']}) 

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.667853
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Analyze the results

Step 3- analyze results
Overall the numbers differ just slightly from the means of the subset dataframe. Leaving the overall pattern to be the same as the subset.

### Create loyalty falg

In [11]:
#step 4- create loyalty flag
#split, transform, and create new column 
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max) 

C:\Users\hardy\AppData\Local\Temp\ipykernel_1788\4099786225.py:3: FutureWarning: The provided callable <function max at 0x0000011BAB742840> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [12]:
#check output
ords_prods_merge.head(10) 

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   
5   2398795        1             2                   3                  7   
6   2398795        1             2                   3                  7   
7   2398795        1             2                   3                  7   
8   2398795        1             2                   3                  7   
9   2398795        1             2                   3                  7   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   
5                    15.0        False         196                  1   
6                    15.0        False       10258                  2   
7                    15.0        False       12427                  3   
8                    15.0        False       13176                  4   
9                    15.0        False       26088                  5   

   reordered  ...                             product_name aisle_id  \
0          0  ...                                     Soda       77   
1          0  ...  Organic Unsweetened Vanilla Almond Milk       91   
2          0  ...                      Original Beef Jerky       23   
3          0  ...               Aged White Cheddar Popcorn       23   
4          0  ...         XL Pick-A-Size Paper Towel Rolls       54   
5          1  ...                                     Soda       77   
6          0  ...                               Pistachios      117   
7          1  ...                      Original Beef Jerky       23   
8          0  ...                   Bag of Organic Bananas       24   
9          1  ...               Aged White Cheddar Popcorn       23   

   department_id  prices  Match    price_range_loc     busiest_day  \
0              7     9.0   both  Mid-range product  Regularly busy   
1             16    12.5   both  Mid-range product  Regularly busy   
2             19     4.4   both  Low-range product  Regularly busy   
3             19     4.7   both  Low-range product  Regularly busy   
4             17     1.0   both  Low-range product  Regularly busy   
5              7     9.0   both  Mid-range product  Regularly busy   
6             19     3.0   both  Low-range product  Regularly busy   
7             19     4.4   both  Low-range product  Regularly busy   
8              4    10.3   both  Mid-range product  Regularly busy   
9             19     4.7   both  Low-range product  Regularly busy   

     busiest_days busiest_period_of_day max_order  
0  Regularly busy        Average orders        10  
1  Regularly busy        Average orders        10  
2  Regularly busy        Average orders        10  
3  Regularly busy        Average orders        10  
4  Regularly busy        Average orders        10  
5    Slowest days        Average orders        10  
6    Slowest days        Average orders        10  
7    Slowest days        Average orders        10  
8    Slowest days        Average orders        10  
9    Slowest days        Average orders        10  

[10 rows x 21 columns]

In [13]:
#check relevant columns 
ords_prods_merge[['user_id', 'order_number', 'max_order']].head(20) 

user_id  order_number  max_order
0         1             1         10
1         1             1         10
2         1             1         10
3         1             1         10
4         1             1         10
5         1             2         10
6         1             2         10
7         1             2         10
8         1             2         10
9         1             2         10
10        1             2         10
11        1             3         10
12        1             3         10
13        1             3         10
14        1             3         10
15        1             3         10
16        1             4         10
17        1             4         10
18        1             4         10
19        1             4         10

#### Question 
Like this- I am not sure why I am only getting info for user_id one and getting multiple repeats of the order number

In [14]:
#create flag filters
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [15]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [16]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [18]:
#check frequency
ords_prods_merge['loyalty_flag'].value_counts(dropna = False) 

loyalty_flag
Regular customer    15874128
Loyal customer      10282763
New customer         6242841
Name: count, dtype: int64

In [19]:
#check sample
ords_prods_merge[['user_id', 'max_order', 'loyalty_flag']].head(25) 

user_id  max_order  loyalty_flag
0         1         10  New customer
1         1         10  New customer
2         1         10  New customer
3         1         10  New customer
4         1         10  New customer
5         1         10  New customer
6         1         10  New customer
7         1         10  New customer
8         1         10  New customer
9         1         10  New customer
10        1         10  New customer
11        1         10  New customer
12        1         10  New customer
13        1         10  New customer
14        1         10  New customer
15        1         10  New customer
16        1         10  New customer
17        1         10  New customer
18        1         10  New customer
19        1         10  New customer
20        1         10  New customer
21        1         10  New customer
22        1         10  New customer
23        1         10  New customer
24        1         10  New customer

#### Group data

In [20]:
#step 5-  The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified
#group data
ords_prods_merge.groupby('loyalty_flag')['prices'].mean() 

loyalty_flag
Loyal customer      7.773575
New customer        7.801206
Regular customer    7.798262
Name: prices, dtype: float64

The mean prices of items grouped by loyalty of customers are about the same at ~$7.80 

### Target different types of spenders

In [22]:
#Step No. 6: The team now wants to target different types of spenders in their marketing campaigns
#create a spending flag for users based on the average price across all their orders
ords_prods_merge['avg_spend_user'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean) 

C:\Users\hardy\AppData\Local\Temp\ipykernel_1788\1853735724.py:3: FutureWarning: The provided callable <function mean at 0x0000011BAB743240> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['avg_spend_user'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [24]:
#transform data
ords_prods_merge['avg_spend_user'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\hardy\AppData\Local\Temp\ipykernel_1788\482631266.py:2: FutureWarning: The provided callable <function mean at 0x0000011BAB743240> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['avg_spend_user'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [25]:
#check output
ords_prods_merge[['user_id', 'prices', 'avg_spend_user']].head(25)

user_id  prices  avg_spend_user
0         1     9.0        6.367797
1         1    12.5        6.367797
2         1     4.4        6.367797
3         1     4.7        6.367797
4         1     1.0        6.367797
5         1     9.0        6.367797
6         1     3.0        6.367797
7         1     4.4        6.367797
8         1    10.3        6.367797
9         1     4.7        6.367797
10        1     4.0        6.367797
11        1     9.0        6.367797
12        1     4.4        6.367797
13        1     3.0        6.367797
14        1     8.6        6.367797
15        1     3.2        6.367797
16        1     9.0        6.367797
17        1     4.4        6.367797
18        1     3.0        6.367797
19        1     8.6        6.367797
20        1     1.0        6.367797
21        1     9.0        6.367797
22        1     4.4        6.367797
23        1     3.0        6.367797
24        1     8.6        6.367797

In [26]:
#set up filter
ords_prods_merge.loc[ords_prods_merge['avg_spend_user'] < 10, 'spending_flag'] = 'Low spender'

In [27]:
#set up filter 
ords_prods_merge.loc[ords_prods_merge['avg_spend_user'] >= 10, 'spending_flag'] = 'High spender'

In [28]:
#check frequency 
ords_prods_merge['spending_flag'].value_counts(dropna = False) 

spending_flag
Low spender     32280013
High spender      119719
Name: count, dtype: int64

In [29]:
#check outputs
ords_prods_merge[['user_id', 'avg_spend_user', 'spending_flag']].head(25) 

user_id  avg_spend_user spending_flag
0         1        6.367797   Low spender
1         1        6.367797   Low spender
2         1        6.367797   Low spender
3         1        6.367797   Low spender
4         1        6.367797   Low spender
5         1        6.367797   Low spender
6         1        6.367797   Low spender
7         1        6.367797   Low spender
8         1        6.367797   Low spender
9         1        6.367797   Low spender
10        1        6.367797   Low spender
11        1        6.367797   Low spender
12        1        6.367797   Low spender
13        1        6.367797   Low spender
14        1        6.367797   Low spender
15        1        6.367797   Low spender
16        1        6.367797   Low spender
17        1        6.367797   Low spender
18        1        6.367797   Low spender
19        1        6.367797   Low spender
20        1        6.367797   Low spender
21        1        6.367797   Low spender
22        1        6.367797   Low spender
23        1        6.367797   Low spender
24        1        6.367797   Low spender

#### Create Frequency flag

In [30]:
#step 7 
#Create an order frequency flag that marks the regularity of a users ordering behavior according to the median in the days_since_prior_order column.
ords_prods_merge['ord_regularity_median'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median) 

C:\Users\hardy\AppData\Local\Temp\ipykernel_1788\3748418766.py:3: FutureWarning: The provided callable <function median at 0x0000011BAB89A7A0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['ord_regularity_median'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [31]:
#check results
ords_prods_merge[['user_id', 'days_since_prior_order', 'ord_regularity_median']].head(25) 

user_id  days_since_prior_order  ord_regularity_median
0         1                     NaN                   20.5
1         1                     NaN                   20.5
2         1                     NaN                   20.5
3         1                     NaN                   20.5
4         1                     NaN                   20.5
5         1                    15.0                   20.5
6         1                    15.0                   20.5
7         1                    15.0                   20.5
8         1                    15.0                   20.5
9         1                    15.0                   20.5
10        1                    15.0                   20.5
11        1                    21.0                   20.5
12        1                    21.0                   20.5
13        1                    21.0                   20.5
14        1                    21.0                   20.5
15        1                    21.0                   20.5
16        1                    29.0                   20.5
17        1                    29.0                   20.5
18        1                    29.0                   20.5
19        1                    29.0                   20.5
20        1                    29.0                   20.5
21        1                    28.0                   20.5
22        1                    28.0                   20.5
23        1                    28.0                   20.5
24        1                    28.0                   20.5

In [32]:
#add filters
ords_prods_merge.loc[ords_prods_merge['ord_regularity_median'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [33]:
ords_prods_merge.loc[(ords_prods_merge['ord_regularity_median'] > 10) & (ords_prods_merge['ord_regularity_median'] <= 20), 'frequency_flag'] = 'Regular customer'

In [34]:
ords_prods_merge.loc[ords_prods_merge['ord_regularity_median'] <= 10, 'frequency_flag'] = 'Frequent customer' 

In [35]:
#check results
ords_prods_merge[['user_id', 'days_since_prior_order', 'ord_regularity_median', 'frequency_flag']].head(25) 

user_id  days_since_prior_order  ord_regularity_median  \
0         1                     NaN                   20.5   
1         1                     NaN                   20.5   
2         1                     NaN                   20.5   
3         1                     NaN                   20.5   
4         1                     NaN                   20.5   
5         1                    15.0                   20.5   
6         1                    15.0                   20.5   
7         1                    15.0                   20.5   
8         1                    15.0                   20.5   
9         1                    15.0                   20.5   
10        1                    15.0                   20.5   
11        1                    21.0                   20.5   
12        1                    21.0                   20.5   
13        1                    21.0                   20.5   
14        1                    21.0                   20.5   
15        1                    21.0                   20.5   
16        1                    29.0                   20.5   
17        1                    29.0                   20.5   
18        1                    29.0                   20.5   
19        1                    29.0                   20.5   
20        1                    29.0                   20.5   
21        1                    28.0                   20.5   
22        1                    28.0                   20.5   
23        1                    28.0                   20.5   
24        1                    28.0                   20.5   

           frequency_flag  
0   Non-frequent customer  
1   Non-frequent customer  
2   Non-frequent customer  
3   Non-frequent customer  
4   Non-frequent customer  
5   Non-frequent customer  
6   Non-frequent customer  
7   Non-frequent customer  
8   Non-frequent customer  
9   Non-frequent customer  
10  Non-frequent customer  
11  Non-frequent customer  
12  Non-frequent customer  
13  Non-frequent customer  
14  Non-frequent customer  
15  Non-frequent customer  
16  Non-frequent customer  
17  Non-frequent customer  
18  Non-frequent customer  
19  Non-frequent customer  
20  Non-frequent customer  
21  Non-frequent customer  
22  Non-frequent customer  
23  Non-frequent customer  
24  Non-frequent customer

In [36]:
#check frequency 
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21556644
Regular customer          7207340
Non-frequent customer     3635743
NaN                             5
Name: count, dtype: int64

### Export data

In [38]:
#export file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_Task_4_8.pkl')) 